Buildings/Fluid/HeatExchangers/Examples/WetCoilCounterFlowPControlAutoTuning.mo

```diff
@@ -0,0 +1,213 @@
+    from_dp=true,
+    redeclare package Medium = Medium1,
+    dp_nominal=3000,
+    m_flow_nominal=m1_flow_nominal)
+    annotation (Placement(transformation(extent={{90,50},{110,70}})));
+  Buildings.Fluid.Sensors.TemperatureTwoPort temSen(redeclare package Medium =
+               Medium2, m_flow_nominal=m2_flow_nominal)
+    annotation (Placement(transformation(extent={{20,10},{0,30}})));
+  Buildings.Fluid.Actuators.Valves.TwoWayEqualPercentage val(
+    redeclare package Medium = Medium1,
+    m_flow_nominal=m1_flow_nominal,
+    dpValve_nominal=6000)
+    "Valve model"
+    annotation (Placement(transformation(extent={{30,50},{50,70}})));
+  Modelica.Blocks.Sources.TimeTable TSet(table=[0,288.15; 600,288.15; 600,
+        298.15; 1200,298.15; 1800,283.15; 2400,283.15; 2400,288.15])
+    "Setpoint temperature" annotation (Placement(transformation(extent={{-80,90},
+            {-60,110}})));
+  Buildings.Fluid.HeatExchangers.WetCoilCounterFlow hex(
+    redeclare package Medium1 = Medium1,
+    redeclare package Medium2 = Medium2,
+    m1_flow_nominal=m1_flow_nominal,
+    m2_flow_nominal=m2_flow_nominal,
+    dp2_nominal(displayUnit="Pa") = 200,
+    allowFlowReversal1=true,
+    allowFlowReversal2=true,
+    dp1_nominal(displayUnit="Pa") = 3000,
+    UA_nominal=2*m1_flow_nominal*4200*(T_a1_nominal - T_b1_nominal)/
+      Buildings.Fluid.HeatExchangers.BaseClasses.lmtd(
+        T_a1_nominal,
+        T_b1_nominal,
+        T_a2_nominal,
+        T_b2_nominal),
+    show_T=true,
+    energyDynamics=Modelica.Fluid.Types.Dynamics.FixedInitial)
+    annotation (Placement(transformation(extent={{60,16},{80,36}})));
+  Modelica.Blocks.Sources.Constant const(k=0.8)
+    annotation (Placement(transformation(extent={{100,-70},{120,-50}})));
+  Buildings.Utilities.Psychrometrics.X_pTphi x_pTphi(use_p_in=false)
+    annotation (Placement(transformation(extent={{150,-42},{170,-22}})));
+  Modelica.Blocks.Sources.Constant const1(k=T_a2_nominal)
+    annotation (Placement(transformation(extent={{100,-38},{120,-18}})));
+  Buildings.Controls.OBC.Utilities.PIDWithAutotuning.FirstOrderAMIGO
+    con(controllerType=Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Types.SimpleController.PI,
+    r=10,
+    yLow=0.2,
+    deaBan=0.2,
+    yRef=0.5,
+    reverseActing=false)
+    "Controller"
+    annotation (Placement(transformation(extent={{0,90},{20,110}})));
+  Modelica.Blocks.Sources.Ramp TWat(
+    height=5,
+    offset=T_a1_nominal,
+    startTime=300,
+    duration=2000) "Water temperature, raised to high value at t=3000 s"
+    annotation (Placement(transformation(extent={{-80,54},{-60,74}})));
+  Buildings.Controls.OBC.CDL.Logical.Sources.Constant resSig(k=false)
+    "Reset signal"
+    annotation (Placement(transformation(extent={{-80,160},{-60,180}})));
+  Buildings.Controls.OBC.CDL.Logical.Sources.Pulse autTunSig(
+    width=0.1,
+    period=4000,
+    shift=400)
+    "Signal for enabling the autotuning"
```

Comment:
If we change the value as 
```
width=0.2,
period=2000,
shift=100
```
The simulation will fail with the error:
```
Integration started at T = 0 using integration method DASSL
(DAE multi-step solver (dassl/dasslrt of Petzold modified by Dassault Systemes))
Error: The following error was detected at time: 2493.834817548029

Model error - Modelica.Math.log (con.conProMod.timConDel.log.u) = Modelica.Math.log (-25.3528)
```

This is the error due to the instance `log` in block `PIDWithAutotuning.SystemIdentification.FirstOrderTimedelayed.BaseClasses.TimeConstantDelay`, as the block cannot have a negative input.

Reply:
Yes, this autotuning process employs an empirical equation:
![image](https://github.com/lbl-srg/modelica-buildings/assets/6975889/a3ae11d7-805e-4d9d-aff8-ad54fc9632f1).
Since multiple variables are involved, avoid this error seems to be difficult.
Should we discuss it in our meeting next week?

Reply:
Based on the discussion on 6/12/2024

- [ ]  Add an assert to check if the setpoint changes during the autotuning process. If the setpoint changes, stop the autotuning. In this case, the autotuning completes unsuccessfully.
- [ ]  Add a Boolean output to indicate if the autotuning completes successfully or not.
- [ ]  If the autotuning completes unsuccessfully, use the gains before the autotuning starts

Reply:
With the updated model, the simulation output is shown below:

![image](https://github.com/lbl-srg/modelica-buildings/assets/6975889/69c24621-f5ae-41ea-85ac-536bdef56750)

![image](https://github.com/lbl-srg/modelica-buildings/assets/6975889/ae2724a7-b9bf-4aa9-a3ac-3e4595c461ee)

Specifically, at time 2100 seconds, the model detects an abnormal change in the setpoint;
As a result, the autotuning fails at ~2493 seconds, no changes are made to the control gains.